### DISCUSSION TITLE
How do I specify job dependency running in another workflow?

### DISCUSSION THREAD

**Discussion by rn4n** [community/community#26632] [Actions]
<p>I’ve this workflow:</p>
<pre><code>name: Testing
on: [push, pull_request]

jobs:
  testing:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v1
      - name: Setup Environment (Using NodeJS 12.x)
        uses: actions/setup-node@v1
        with:
          node-version: 12.x

      - name: install dependencies
        run: npm install

      - name: Run unit tests and collect coverage
        run: npm run test:unit
</code></pre>
<p>Now, I would like to create another workflow (deploy.yml) to deploy my application on Netlify.</p>
<p>How can specify that this deploy depends of the successfully run of “testing” job ?</p>
<p>“testing” and “deploy” are in different workflows files.</p>
<pre><code>name: Deploy
on:
  push:
    branches:
      - master
jobs:
deploy:
# ... ??? ...
</code></pre>
<p>Thanks.</p>

**Comment by brpaz** [comment, marked as answer]
<p>You cant.</p>
<p>What you can do is to use the same workflow file and then use conditions to trigger or not a specific job.</p>
<p>If you want to run a job only when there is a push to master branch you can do it like this:</p>
<pre><code>jobs:
   deploy:
       if: github.event_name == 'push' &amp;&amp; github.ref == 'refs/heads/master'
</code></pre>
<p>Another option for deployments is to create a tag when you are ready to deploy and then having a seperate workflow that is listening to the push tag event and trigger your deploy.</p>

**Reply by neopunisher** [reply]
I think you can do it now with workflow run https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_run described here https://stackoverflow.com/questions/58457140/dependencies-between-workflows-on-github-actions

**Reply by daveisfera** [reply]
To anyone looking to use this, it only runs on the default branch and not the branch that actually triggered the workflow that this one depends on ( see notes about this [here](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_run) )

**Comment by justfathi** [comment]
<p>Will this be a feature that will be added in the future?</p>

**Comment by mingrammer** [comment]
<p>I think it should be supported!</p>

**Comment by DannyBen** [comment]
<p>I would love to have such a feature. Perhaps it is simple (for GitHub) to implement, by adding another “on” event trigger?</p>
<pre><code>on:
  workflow_complete:
workflow: Test
    types: [success]

jobs:
  ...
</code></pre>
<p>or</p>
<pre><code>on:
  workflow_success:
    workflow: Test
</code></pre>
<p>Seems like a must have for “test then deploy” flow, where it makes sense to separate the two workflows instead of toying around with conditions and have everything as one pile of YAML mess.</p>

**Comment by moltar** [comment]
<p>This solution would cause a problem if you are using matrix. It will run for every matrix item basically.</p>
<p>This really needs to be solved. I can’t believe this doesn’t work!</p>
<p>It’s a pretty common requirement.</p>
<ol>
<li>I need to run tests in a matrix.</li>
</ol>
<p> 2. I also want to deploy only on master, but only if all tests pass.</p>
<p>Right now this is basically impossible.</p>

**Comment by lulubobst** [comment]
<p>When will this be supported?</p>

**Comment by shinny-hongyan** [comment]
<p>I have try this code,</p>
<pre><code>on:
  workflow_success:
    workflow: Test
</code></pre>
<p>but it does not work , and show me</p>
<pre><code>`workflow_success` is not a valid event name
</code></pre>

**Comment by slinkydeveloper** [comment]
<p>Same here, I’m facing the same problem. On every master push I need to test different JDK versions and then I publish a package snapshot, but i don’t know how to solve it.</p>

**Comment by jeffmaher** [comment]
<p>Instead of creating a new workflow, you might just create a dependent job. See the documentation for <a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idneeds" rel="nofollow noopener"><code>jobs.&lt;job_id&gt;.needs</code></a>.</p>

**Comment by AlexMcConnell** [comment]
<p>Yes, you <em>can</em> put it all into different jobs in one workflow, but breaking up your large files into logically separated smaller files is a pretty universal programming best practice, and that’s extra true when you have a lot of different things branching off from one spot (like your tests passing or whatever). If they want to define a workflow as all of those branches together, that’s fine, but that means we need to be able to split out jobs into different files. Otherwise, we’re forced to either write code in some scripting language that builds into the yaml code for our workflow or create some sort of funky trigger event, both of which are incredibly hacky.</p>

**Comment by asarkar** [comment]
<p>This is possible using the <a href="https://docs.github.com/en/rest/reference/actions#workflow-runs" rel="nofollow noopener">GitHub Workflow runs API</a>. Assuming <code>Testing</code> workflow fails if the <code>testing</code> step fails, in the <code>Deploy</code> workflow, you get the conclusion of the latest <code>Testing</code> workflow run using some cURL and jq foo, and then exit if the output is not <code>success</code>. See <a href="https://github.com/asarkar/build-time-tracker/tree/master/.github/workflows" rel="nofollow noopener">this project</a>, where <code>publish</code> depends on <code>ci</code>.</p>

**Comment by NasAmin** [comment]
<p>I am also interested in this as well. Workflow depencencies is a must to allow engineers breakdown their pipelines logically in separate files instead of having to maintain huge files.<br>
I thought I didn’t like GitLab’s approach but they are miles ahead!</p>

**Comment by edudepetris** [comment]
<p>could be possible with <code>workflow_run</code>…</p>
<p>Ref.: <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#workflow_run" class="inline-onebox" rel="noopener nofollow ugc">Events that trigger workflows - GitHub Docs</a></p>
<aside class="quote quote-modified" data-post="1" data-topic="128001">
  <div class="title">
    <div class="quote-controls"></div>
    
    <a href="https://github.community/t/workflow-run-completed-event-triggered-by-failed-workflow/128001">Workflow_run completed event triggered by failed workflow</a> <a class="badge-wrapper  bullet" href="/c/code-to-cloud/github-actions/41"><span class="badge-category-parent-bg" style="background-color: #6a737d;"></span><span class="badge-category-bg" style="background-color: #2188ff;"></span><span style="" data-drop-close="true" class="badge-category clear-badge" title="Discussion and questions about GitHub Actions—automation for all of your development workflows, now with world-class CI/CD.">GitHub Actions</span></a>
  </div>
  <blockquote>
    I have a workflow triggered as follows: 
name: Production deploy
on:
  workflow_run:
    workflows: ["Stage deploy"]
    types:
      - completed

I expect that after the stage deploy (which also runs API tests) completes successfully, then the production deploy would start. However, I just watched a failed stage deploy (which skipped its API tests due to the failed deploy) also trigger a production deploy. Is there a different “types” value for workflow_run that causes it not to fire if the wor…
  </blockquote>
</aside>

**Comment by mobeigi** [comment]
<p>Surprised this isn’t supported. It would be nice to have multiple workflows nicely separated (one for builds, one for tests, one for deployments) and have some depend on others. It’s possible now using all jobs in a single workflow but its a little messier imo.</p>

**Comment by jakoch** [comment]
<p>Hmmm… so.<br>
You go from “build” workflow to “API” to “status var” to “check, if you can publish” to “publish”.<br>
This is a really clever workaround for a missing feature!</p>
